extension/basicauthextension/extension.go

```diff
@@ -0,0 +1,160 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package basicauthextension // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/basicauthextension"
+
+import (
+	"context"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/tg123/go-htpasswd"
+	"go.opentelemetry.io/collector/client"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config/configauth"
+)
+
+var (
+	errNoCredentialSource  = errors.New("no credential source provided")
+	errNoAuth              = errors.New("no basic auth provided")
+	errInvalidCredentials  = errors.New("invalid credentials")
+	errInvalidSchemePrefix = errors.New("invalid authorization scheme prefix")
+	errInvalidFormat       = errors.New("invalid authorization format")
+)
+
+type basicAuth struct {
+	htpasswd  HtpasswdSettings
+	matchFunc func(username, password string) bool
+}
+
+func newExtension(cfg *Config) (configauth.ServerAuthenticator, error) {
+	if cfg.Htpasswd.File == "" && cfg.Htpasswd.Inline == "" {
+		return nil, errNoCredentialSource
+	}
+	ba := basicAuth{
+		htpasswd: cfg.Htpasswd,
+	}
+	return configauth.NewServerAuthenticator(configauth.WithStart(ba.start), configauth.WithAuthenticate(ba.authenticate)), nil
+}
+
+func (ba *basicAuth) start(ctx context.Context, host component.Host) error {
+	var rs []io.Reader
+
+	if ba.htpasswd.File != "" {
+		f, err := os.Open(ba.htpasswd.File)
+		if err != nil {
+			return fmt.Errorf("open htpasswd file: %w", err)
+		}
+		defer f.Close()
+
+		rs = append(rs, f)
+		rs = append(rs, strings.NewReader("\n"))
+	}
+
+	// Ensure that the inline content is read the last.
+	// This way the inline content will override the content from file.
+	rs = append(rs, strings.NewReader(ba.htpasswd.Inline))
+	mr := io.MultiReader(rs...)
+
+	htp, err := htpasswd.NewFromReader(mr, htpasswd.DefaultSystems, nil)
+	if err != nil {
+		return fmt.Errorf("read htpasswd content: %w", err)
+	}
+
+	ba.matchFunc = htp.Match
+
+	return nil
+}
+
+const (
+	metadataKey        = "authorization"
```

Comment:
HTTP headers are case-insensitive, so, the best solution is to iterate over all the headers and use something like strings.EqualFold to check the header name. You can keep those two to short-circuit the checks, like, check first the canonical, then the metadata, then iterate over the headers and do an equal fold.

Reply:
Yeah, makes sense. I assumed they will be either canonicalized or lower-cased as coming either from `confighttp` or `configgrpc`.